File: LeetCode/Medium/1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.c
// Check if the path can form a palindrome
bool canBePalindrome(int *counter) {
    int oddCount = 0;
    for (int i = 0; i < 10; ++i) {
        if (counter[i] % 2 != 0) {
            oddCount++;
        }
    }
    return oddCount <= 1;
}

// Depth-first search to count pseudo-palindromic paths
int dfs(struct TreeNode* node, int *counter) {
    if (!node) return 0;
    
    // Update the frequency counter for the current node's value
    counter[node->val]++;
    
    int result = 0;
    // Check if it's a leaf node
    if (!node->left && !node->right) {
        // Check if the path is pseudo-palindromic
        result = canBePalindrome(counter) ? 1 : 0;
    } else {
        // Continue DFS on left and right children
        result = dfs(node->left, counter) + dfs(node->right, counter);
    }
    
    // Backtrack - undo the frequency counter update
    counter[node->val]--;
    
    return result;
}

// Main function to be called with the root of the tree
int pseudoPalindromicPaths(struct TreeNode* root) {
    int counter[10] = {0}; // Assuming values are digits 0-9
    return dfs(root, counter);
}

File: LeetCode/Medium/1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.py
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def canBePalindrome(self, counter):
        odd_count = sum(v % 2 for v in counter.values())
        return odd_count <= 1
        
    def dfs(self, node, counter):
        if not node:
            return 0
        
        # Update the frequency map for the current node's value
        counter[node.val] = counter.get(node.val, 0) + 1
        
        # If it's a leaf node, check if the path is pseudo-palindromic
        if not node.left and not node.right:
            result = 1 if self.canBePalindrome(counter) else 0
        else:
            # Continue DFS on left and right children
            result = self.dfs(node.left, counter) + self.dfs(node.right, counter)
        
        # Backtrack: undo the frequency map update before returning
        counter[node.val] -= 1
        
        return result

    def pseudoPalindromicPaths (self, root: TreeNode) -> int:
        
        # Start DFS from the root with an empty frequency map
        return self.dfs(root, {})
